### ISSUE TITLE
ColorConstants fields are initialized to null in a Papyrus context

### ISSUE THREAD

**Issue by VincentLorenzo** [eclipse/gef-classic#201]
Papyrus worked fine with the previous GEF version (Eclipse 2023-03). 
Now, we are moving on the last version of GEF for 2023-06 and this version is breaking Papyrus Diagram editors. 

After investigation, the problem comes from the changes done in ColorConstants. 

Previously the constants like 'button' were initialized using the Display -> no problem for us. 
Now the constants are initialized using the Eclipse Theme mechanism. Unfortunately for Papyrus, we call ColorConstants before the initialization of the Eclipse Theme. 

Consequently, all fields from ColorConstants initialized with a call to ColorProvider.SystemColorFactory.colorProvider.getXXX are set to null.

**Comment by azoitl** [comment]
Just to give you some background the change in question is needed so that GEF finally can better adhere to the current Eclipse Platform theme (e..g, dark theme). We've tested that in different settings, also an RCP similar to Papyrus. Also with the palette. Because the palette was a major reason for the update. 

That aside, we need to find a fix for your problem. Is there an easy way that I can debug papyrus on my machine? Do you know by accident know which part is accessing the GEF colors so early during Eclipse platform setup that the theme is not ready? Because the trace you provide is the symptom but unfortunately not the cause.

**Comment by VincentLorenzo** [comment]
Sorry, I forgot to mention that I also filled a Papyrus bug with more details explaining how to reproduce the problem https://bugs.eclipse.org/bugs/show_bug.cgi?id=581906 . 

From an Eclipse Commiter 2023-06-M1: 
you can install Papyrus from the nightly build update site : https://download.eclipse.org/modeling/mdt/papyrus/updates/nightly/master/ . This action will install these draw2D and GEF versions: 
Eclipse GEF	GEF Classic Draw2d	3.13.0.202305031354	org.eclipse.draw2d
Eclipse GEF	GEF Classic GEF (MVC)	3.14.0.202305031354	org.eclipse.gef

After restarting Eclipse, You can create a new Papyrus Project, with a new Model and a Class Diagram to get the error. 

My investigations:
I found 3 Papyrus Diagram plugins contributing to the extension point **org.eclipse.ui.services** with a _sourceProvide_ . These plugins are defined in their Manifest.MF with **Bundle-ActivationPolicy: lazy** and **Eclipse-LazyStart: true**, so when Eclipse load the extension point **org.eclipse.ui.services**, their Activator (class _UMLDiagramEditorPlugin_) of these plugins are loaded and started. 
Their activator methods **UMLDiagramEditorPlugin#start** call 
- **DiagramPreferenceInitializer.initializeDefaultPreferences()** 
- which calls **org.eclipse.gmf.runtime.diagram.ui.preferences.AppearancePreferencePage.initDefaults()** 
- which call **org.eclipse.draw2d.ColorConstants.black.**. 
The black color is well resolved, but this call also initializes all the constants of ColorConstants and the constants defining with **ColorProvider.SystemColorFactory.xxx** are set to null. 

Here the 3 diagram plugins name I found:
org.eclipse.papyrus.uml.diagram.component
org.eclipse.papyrus.uml.diagram.composite
org.eclipse.papyrus.uml.diagram.deployment

FYI:
1. Papyrus is moving to java 17 with this patch https://git.eclipse.org/r/c/papyrus/org.eclipse.papyrus/+/201723/3
1.1 java 17
1.2 update targetplatform for Eclipse 2023-06, excepted GEF
2. Papyrus migrates its code to be compliant with new GEF methods signature : https://git.eclipse.org/r/c/papyrus/org.eclipse.papyrus/+/201721 (chained to the previous one)
2.1 update GEF version in the target platform
2.2 fix GEF API usages in papyrus
3. I tried a **temporary** patch to fix the current issue on the Papyrus side https://git.eclipse.org/r/c/papyrus/org.eclipse.papyrus/+/201755/1 (also chained with the previous ones), but we would prefer to find another solution
3.1 it works locally
3.2 as for previous patch, Jenkins seems to have difficulties to build it...

**Comment by azoitl** [comment]
Thanks for the detailed analysis. This helps a lot to understand the problem. So how I see it is happening how GMF runtime is setting up their preferences. This is something I do not understand to well, but maybe should be reconsidered in the long run. Activators are considered a startup performance issues and should be minimized as much as possible. 

However I think this is nothing quick. What I could offer as a quick fix from Eclipse GEF Classic side is to implement a fall back: when the theme is not providing a color we could give you back the color that was used before. The only drawback would be that  then theming would not work. 

Regarding your other problems: if I can be of any help. Let me know I had a look on your patch for the methods. I would have some suggestions to further reduce the code, making you more resilient to future changes. 

I saw at least one of your Jenkins builds that maven is running out of heap space. We had the same issue with Eclipse 4diac builds. We solved that by adding the -Xmx2048m to the java options of maven. I hope this helps.

**Comment by pcdavid** [comment]
@VincentLorenzo could you test with the GMF Runtime at https://download.eclipse.org/modeling/gmp/gmf-runtime/updates/milestones/S202305070953/ ?
I tried removing the dependency from GMF's `AppearancePreferencePage` to GEF's `ColorConstants`.

**Comment by VincentLorenzo** [comment]
@pcdavid , I thank you for your fix. I'm sorry, but I still get the issue with your update site (Fresh Eclipse Installation + your update site + Papyrus), but it works fine from the source code, taking the plugin org.eclipse.gmf.runtime.diagram.ui in my workspace and launching a runtime. 

@azoitl , I thank you for your help. Concerning Jenkins, we probably have a dialog opens blocking our tests or a new issue with our html renderer in the diagrams. I will investigate in this direction today.

**Comment by pcdavid** [comment]
Yeah, I must have triggered the build too quickly, it seems the content of the p2 repo corresponds to the previous commit. 
Can you try with https://download.eclipse.org/modeling/gmp/gmf-runtime/updates/milestones/S202305091320/ ?

**Comment by VincentLorenzo** [comment]
@pcdavid, I confirm the new GMF-Runtime build fix this issue for Papyrus. Thank you for your help.

**Comment by azoitl** [comment]
@VincentLorenzo @pcdavid  happy to hear that. I hope with that also improved theming support will be available for papyrus and other GMF/Sirus based editors. 

Just that you know one cause for the problem is that in the in the eclipse platform the theming engine is loaded to late. Some details can be found in this issue https://github.com/eclipse-platform/eclipse.platform.ui/issues/259 especially my comments.

Can we close this issue then?

**Comment by VincentLorenzo** [comment]
@azoitl Yes, you can close the issue. 

As I'm not familiar with Eclipse Theme, could you indicate me if there is an obvious way to check we get benefits on a Papyrus diagram with the changes done on GEF side ? On my side, in the Dark mode, my palette still have the standart color scheme (grey/white), but as Papyrus overrides many things, maybe it is due to Papyrus. I should also check with Sirius asap.

**Comment by azoitl** [comment]
@VincentLorenzo yes you should see improvements in the palette, the ruler and depending on the implementation in the editor backgrounds. After switching to dark mode you need to restart. The current design of GEF Classic does not allow to update colors dynamically and we didn't want to break more.
